diff --git a/panels/display/cc-display-panel.c b/panels/display/cc-display-panel.c
index f9d4ae044..ef2a8a693 100644
--- a/panels/display/cc-display-panel.c
+++ b/panels/display/cc-display-panel.c
@@ -853,7 +853,7 @@ rebuild_ui (CcDisplayPanel *panel)
   n_outputs = g_list_length (outputs);
   type = config_get_current_type (panel);
 
-  if (n_outputs == 2 && n_usable_outputs == 2)
+  if (n_outputs == 2 && n_usable_outputs == 99)
     {
       /* We only show the top chooser with two monitors that are
        * both usable (i.e. two monitors incl. internal and lid not closed).
